scala-package/core/src/main/scala/ml/dmlc/mxnet/io/NDArrayIter.scala

```diff
@@ -38,15 +38,14 @@ import scala.collection.immutable.ListMap
  * the size of data does not match batch_size. Roll over is intended
  * for training and can cause problems if used for prediction.
  */
-class NDArrayIter (data: IndexedSeq[NDArray], label: IndexedSeq[NDArray] = IndexedSeq.empty,
-                  private val dataBatchSize: Int = 1, shuffle: Boolean = false,
-                  lastBatchHandle: String = "pad",
-                  dataName: String = "data", labelName: String = "label") extends DataIter {
+class NDArrayIter (data: IndexedSeq[(String, NDArray)], label: IndexedSeq[(String, NDArray)],
+                  private val dataBatchSize: Int, shuffle: Boolean,
+                  lastBatchHandle: String) extends DataIter {
```

Comment:
I agree with other reviewers that this should be a different constructor and then common constructor code moved to a function and called by both.

Reply:
As far as I know, the [auxiliary constructor](https://www.artima.com/pins1ed/functional-objects.html#6.7) in Scala is not able to call a function other than the primary constructor, hence it seems not possible to move the common constructor code into a function and call it by both constructors?

Reply:
So then can it be structured in such a way that the primary constructor serves as the "common" portion, passing it a translation (of sorts) of the multiple input types?

Reading that, I didn't see where it said it can;t call other functions, but I'll take your word for it.

Reply:
It said:

> In Scala, every auxiliary constructor must invoke another constructor of the same class as its first action. In other words, the first statement in every auxiliary constructor in every Scala class will have the form "this(...)".

So my word was not precise, it can also call other constructors, but still not other functions.

I think this PR already tried to structure the primary constructor as a "common" protion, and use the `def this()` at [line 77](https://github.com/parallelgithub/incubator-mxnet/blob/3bdb52c25b959fd34fac146fe7d2d818b8be6eec/scala-package/core/src/main/scala/ml/dmlc/mxnet/io/NDArrayIter.scala#L77) to translate the old constructor's inputs into the primary constructor to support backward compatibility?

I'm not sure if this will cause any error for the users using current version, though. (At least the test passed :P) Maybe we can examine the test (or add some more tests) to show that it still works in the original use case?

Reply:
@cjolivier01

Thank you for giving a direction to think about it.

Let me explain why it is difficult to design it in that way.
Here is an example of overloading constructor in Scala:

```scala
class A(i: Int) {
  \\ Part 1
  val initData = 5
  def this(i: Int, j: Int) = {
    \\ You can't write anything here
    this(i + j)
    \\ Part 3
  }
  \\ Part 2
  def getData() = initData
}
```
If `new A(10)`, Part 1 and Part 2 will be executed.
If `new A(10, 20)`, Part 1, Part 2, and the Part 3 will be executed.
When you initialize fields and methods in Part 1 and Part 2, it's late to do anything in Part 3.
When executing 'common' portion Part 1 and Part 2, we must pass the original version of argument set, which loses the information of custom names for multiple input.

One possible way is declaring the field as `var`.
Since using mutable variable is not good for maintaining program, I don't recommend this way.

Maybe there is an [alternative strategy](https://github.com/Ldpe2G/DeepLearningForFun/blob/master/Mxnet-Scala/MatrixFactorization/matrixFac.scala) to manipulate multiple input. 
It doesn't use `Module` so that avoid using `NDArrayIter`. 
Since `Module` aims to provide high-level API, I think it's worthy to dedicated to `NDArrayIter`.